wandb/sdk/service/server.py

```diff
@@ -0,0 +1,127 @@
+from .server_sock import SocketServer
+from .streams import StreamMux
+
+
+class WandbServer:
+    _pid: Optional[int]
+    _grpc_port: Optional[int]
+    _sock_port: Optional[int]
+    _debug: bool
+    _serve_grpc: bool
+    _serve_sock: bool
+    _sock_server: Optional[SocketServer]
+
+    def __init__(
+        self,
+        grpc_port: int = None,
+        sock_port: int = None,
+        port_fname: str = None,
+        address: str = None,
+        pid: int = None,
+        debug: bool = False,
+        serve_grpc: bool = False,
+        serve_sock: bool = False,
+    ) -> None:
+        self._grpc_port = grpc_port
+        self._sock_port = sock_port
+        self._port_fname = port_fname
+        self._address = address
+        self._pid = pid
+        self._debug = debug
+        self._serve_grpc = serve_grpc
+        self._serve_sock = serve_sock
+        self._sock_server = None
+
+        if grpc_port:
+            _ = wandb.util.get_module(  # type: ignore
+                "grpc",
+                required="grpc port requires the grpcio library, run pip install wandb[grpc]",
+            )
+
+        debug = True
+        if debug:
+            logging.basicConfig(stream=sys.stdout, level=logging.DEBUG)
+
+    def _inform_used_ports(
+        self, grpc_port: Optional[int], sock_port: Optional[int]
+    ) -> None:
+        if not self._port_fname:
+            return
+        pf = port_file.PortFile(grpc_port=grpc_port, sock_port=sock_port)
+        pf.write(self._port_fname)
+
+    def _start_grpc(self, mux: StreamMux) -> int:
+        from .server_grpc import WandbServicer
+
+        address: str = self._address or "127.0.0.1"
+        port: int = self._grpc_port or 0
+        pid: int = self._pid or 0
+        server = grpc.server(
+            futures.ThreadPoolExecutor(max_workers=10, thread_name_prefix="GrpcPoolThr")
+        )
+        servicer = WandbServicer(server=server, mux=mux)
+        try:
+            spb_grpc.add_InternalServiceServicer_to_server(servicer, server)
+            port = server.add_insecure_port(f"{address}:{port}")
+            mux.set_port(port)
+            mux.set_pid(pid)
+            server.start()
+        except KeyboardInterrupt:
+            mux.cleanup()
+            server.stop(0)
+            raise
+        except Exception:
+            mux.cleanup()
+            server.stop(0)
+            raise
+        return port
+
+    def _start_sock(self, mux: StreamMux) -> int:
+        address: str = self._address or "127.0.0.1"
+        port: int = self._sock_port or 0
+        # pid: int = self._pid or 0
+        self._sock_server = SocketServer(mux=mux, address=address, port=port)
+        try:
+            self._sock_server.start()
+            port = self._sock_server.port
+        except KeyboardInterrupt:
+            mux.cleanup()
+            raise
+        except Exception:
+            mux.cleanup()
+            raise
+        return port
+
+    def _stop_servers(self) -> None:
```

Comment:
do you need to stop the gRPC server?

Reply:
We stop it inband in server_grpc.py:
```
    def ServerShutdown(  # noqa: N802
        self, request: spb.ServerShutdownRequest, context: grpc.ServicerContext,
    ) -> spb.ServerShutdownResponse:
        result = spb.ServerShutdownResponse()
        self._server.stop(5)
        return result
```
( i think this introduces a nasty delay so it could be improved when we want to focus on grpc startup/shutdown performance)